### ISSUE TITLE
[Discussion] Is it better to separate metric and meter?

### ISSUE THREAD

**Issue by Lyken17** [pytorch/ignite#332]
Hi there

While I am trying to write an ImageNet training scripts as [official pytorch imagenet](https://github.com/pytorch/examples/blob/master/imagenet/main.py), I found the `ignite.metrics.TopKCategoricalAccuracy` automatically averages the results, which makes it hard to get the accuracy of current feed-forward. 
```python
    def update(self, output):
        # ...
        self._num_correct += torch.sum(correct).item()
        self._num_examples += correct.shape[0]

    def compute(self):
        if self._num_examples == 0:
            raise NotComputableError('TopKCategoricalAccuracy must have at least one example before it can be computed')
        return self._num_correct / self._num_examples
```

So I am thinking, is it better to separate `metric` and `meter`? The metric only measures the results, while `meter` handles average, momentum moving,  or confusion matrix (like [pytorch/tnt](https://github.com/pytorch/tnt/tree/master/torchnet/meter) did). This will give user flexibility to fetch current measurements, as well as switching between normal average and moving average.

**Comment by vfdev-5** [comment]
@Lyken17 thanks for the feedback! If I understand correctly your problem is that you would like to compute top-k categorical accuracy at each iteration and display it somehow during the training ?

We had a similar issue #296 but we solved it only for outputs and not metrics.
Could you please provide more details on what you mean by metric/meter separation and how it is done in tnt ?

**Comment by Lyken17** [comment]
@vfdev-5 Here is an example code. Hope this will better explain my proposal

```python
class MovingAverageValueMeter(Meter):
	def __init__(self, windowsize):
		super(MovingAverageValueMeter, self).__init__()
		self.windowsize = windowsize
		self.valuequeue = torch.Tensor(windowsize)
		self.reset()

	def reset(self):
		self.sum = 0.0
		self.n = 0
		self.var = 0.0
		self.valuequeue.fill_(0)

	def add(self, value):
		queueid = (self.n % self.windowsize)
		oldvalue = self.valuequeue[queueid]
		self.sum += value - oldvalue
		self.var += value * value - oldvalue * oldvalue
		self.valuequeue[queueid] = value
		self.n += 1

	def value(self):
		n = min(self.n, self.windowsize)
		mean = self.sum / max(1, n)
		std = math.sqrt(max((self.var - n * mean * mean) / max(1, n - 1), 0))
		return mean, std


class AverageMeter(Meter):
	"""code from TNT"""

	def __init__(self):
		self.reset()

	def reset(self):
		self.val = 0
		self.avg = 0
		self.sum = 0
		self.count = 0

	def update(self, val, n=1):
		self.val = val
		self.sum += val * n
		self.count += n
		self.avg = self.sum / self.count


class TopkAcc(Metric):
	def __init__(self, k):
		self.k = k

	def __call__(self, y_pred, y):
		k = self.k
		sorted_indices = torch.topk(y_pred, k, dim=1)[1]
		expanded_y = y.view(-1, 1).expand(-1, k)
		correct = torch.sum(torch.eq(sorted_indices, expanded_y), dim=1)
		return torch.sum(correct).item() / float(correct.shape[0]) * 100


evaluator = create_supervised_evaluator(model,
	metrics={
		'top-1': AverageMeter(TopkAcc(1)), # For accuracy, we want to see the averages
                'top-5': AverageMeter(TopkAcc(1)),
		'nll': MovingAverageValueMeter(Loss()) # For loss, we want observe the moving average
	},
	device=device)

# hooks for evaluator
@evaluator.on(Events.ITERATION_COMPLETED)
def logged_with_smooth(engine):
	metrics = engine.metrics
	top1 = metrics["top-1"]
	top5 = metrics["top-5"]
	loss = metrics["loss"]
	print('Test: [{0}/{1}]\t'
	      'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
	      'Acc@1 {top1.val:.3f} ({top1.avg:.3f})\t'
	      'Acc@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
			i, len(val_loader), loss=loss,
			top1=top1, top5=top5))

```

**Comment by Lyken17** [comment]
The class `Metric`, only measures the model performance of current feed sample.

The class `Meter`, handles update / sum / average  and other stuffs.

**Comment by vfdev-5** [comment]
Okay, thanks, I see better the separation you propose. Could you please explain how this separation could solve your initial problem with logging TopKAccuracy (during training) at each iteration ?

If you would like to have a running average of a metric, as in the example, we have [RunningAverage](https://github.com/pytorch/ignite/blob/master/ignite/metrics/running_average.py#L5).

cc @jasonkriss

**Comment by jasonkriss** [comment]
@Lyken17 I think I understand the proposal but I'm not sure I completely understand the motivation. I think ignite's metric's are actually more or less the same thing as tnt's meter's. E.g. https://github.com/pytorch/tnt/blob/master/torchnet/meter/classerrormeter.py. I think we could add more higher average metric's like `RunningAverage` if people found that useful.

Can you give use a lil more background on why you might want to have raw TopKCategoricalAccuracy of a single iteration?

**Comment by vfdev-5** [comment]
@Lyken17 I'll close the issue as no more activity on this. Feel free to reopen if needed
